File: src/tests/python_tests/physics_animation_tests.py
"""
Copyright (c) 2017 Doyub Kim

I am making my contributions/submissions to this project solely in my personal
capacity and am not conveying any rights to any intellectual property of any
third parties.
"""

import math
import pyjet
import unittest


class MyPhysicsAnimation(pyjet.PhysicsAnimation):
    def __init__(self):
        self.init_data = 0
        self.adv_data = 0
        super(MyPhysicsAnimation, self).__init__()

    def onAdvanceTimeStep(self, timeIntervalInSeconds):
        self.adv_data += 1

    def numberOfSubTimeSteps(self, timeIntervalInSeconds):
        return int(math.ceil(timeIntervalInSeconds / 0.02))

    def onInitialize(self):
        self.init_data = 1


class PhysicsAnimationTests(unittest.TestCase):
    def testInheritance(self):
        anim = MyPhysicsAnimation()

        anim.isUsingFixedSubTimeSteps = False
        f = pyjet.Frame(index=3, timeIntervalInSeconds=0.1)
        anim.update(f)
        self.assertEqual(anim.init_data, 1)
        self.assertEqual(anim.adv_data, 20)


def main():
    pyjet.Logging.mute()
    unittest.main()


if __name__ == '__main__':
    main()

File: src/tests/unit_tests/physics_animation_tests.cpp
// Copyright (c) 2018 Doyub Kim
//
// I am making my contributions/submissions to this project solely in my
// personal capacity and am not conveying any rights to any intellectual
// property of any third parties.

#include <jet/physics_animation.h>

#include <gtest/gtest.h>

using namespace jet;

class CustomPhysicsAnimation : public PhysicsAnimation {
 public:
 protected:
    void onAdvanceTimeStep(double timeIntervalInSeconds) override {
        (void)timeIntervalInSeconds;
    }
};

TEST(PhysicsAnimation, Constructors) {
    CustomPhysicsAnimation pa;
    EXPECT_EQ(-1, pa.currentFrame().index);
}

TEST(PhysicsAnimation, Properties) {
    CustomPhysicsAnimation pa;

    pa.setIsUsingFixedSubTimeSteps(true);
    EXPECT_TRUE(pa.isUsingFixedSubTimeSteps());
    pa.setIsUsingFixedSubTimeSteps(false);
    EXPECT_FALSE(pa.isUsingFixedSubTimeSteps());

    pa.setNumberOfFixedSubTimeSteps(42);
    EXPECT_EQ(42u, pa.numberOfFixedSubTimeSteps());

    pa.setCurrentFrame(Frame(8, 0.01));
    EXPECT_EQ(8, pa.currentFrame().index);
    EXPECT_EQ(0.01, pa.currentFrame().timeIntervalInSeconds);
}

TEST(PhysicsAnimation, Updates) {
    CustomPhysicsAnimation pa;

    for (Frame frame(0, 0.1); frame.index <= 15; ++frame) {
        pa.update(frame);
    }

    EXPECT_DOUBLE_EQ(1.5, pa.currentTimeInSeconds());

    CustomPhysicsAnimation pa2;

    for (int i = 0; i <= 15; ++i) {
        pa2.advanceSingleFrame();
    }

    EXPECT_DOUBLE_EQ(pa2.currentFrame().timeIntervalInSeconds * 15.0,
                     pa2.currentTimeInSeconds());
}